File: src/Runtime/Scala/Runtime.scala
package lego.runtime

import scala.collection.mutable

/**
 * Lego Runtime Library for Scala
 *
 * This module provides the core runtime infrastructure for Lego-generated Scala code.
 * It includes:
 * - Core types (Term, GrammarExpr, etc.)
 * - Parsing engine (parseGrammar, lexGrammar)
 * - IO operations (readFile, writeFile)
 * - Memoization (HashMap-based Packrat)
 *
 * Generated Rosetta code should: import lego.runtime._
 */
object Runtime {

  //============================================================================
  // Core Types
  //============================================================================

  /** The universal Term type for AST representation */
  sealed trait Term
  case class Var(name: String) extends Term
  case class Lit(value: String) extends Term
  case class Con(name: String, args: List[Term]) extends Term

  object Term {
    def show(t: Term): String = t match {
      case Var(n) => s"Var($n)"
      case Lit(v) => s"Lit($v)"
      case Con(n, args) => s"Con($n, [${args.map(show).mkString(", ")}])"
    }
  }

  /** Grammar expression algebra */
  sealed trait GrammarExpr
  case object Empty extends GrammarExpr
  case class GLit(s: String) extends GrammarExpr
  case class Ref(name: String) extends GrammarExpr
  case class Seq(g1: GrammarExpr, g2: GrammarExpr) extends GrammarExpr
  case class Alt(g1: GrammarExpr, g2: GrammarExpr) extends GrammarExpr
  case class Star(g: GrammarExpr) extends GrammarExpr
  case class Plus(g: GrammarExpr) extends GrammarExpr
  case class Opt(g: GrammarExpr) extends GrammarExpr
  case class Node(name: String, g: GrammarExpr) extends GrammarExpr

  /** Grammar production: (name, grammar, annotation) */
  case class Production(name: String, grammar: GrammarExpr, annotation: String = "")

  type Productions = List[Production]

  /** Rewrite rule: name, lhs pattern, rhs template */
  case class Rule(name: String, lhs: Term, rhs: Term)

  //============================================================================
  // Token Types
  //============================================================================

  /** Token types for lexing */
  sealed trait Token
  case class TokLit(value: String) extends Token
  case class TokIdent(name: String) extends Token
  case class TokSym(sym: String) extends Token
  case class TokNumber(num: String) extends Token

  object Token {
    def show(t: Token): String = t match {
      case TokLit(v) => s"\"$v\""
      case TokIdent(n) => n
      case TokSym(s) => s
      case TokNumber(n) => n
    }
  }

  //============================================================================
  // Parse State
  //============================================================================

  /** Packrat memoization entry */
  case class MemoEntry(result: Option[(Term, Int, List[(String, Term)])])

  /** Parse state */
  case class ParseState(
    tokens: List[Token],
    pos: Int = 0,
    memo: mutable.HashMap[(Int, String), MemoEntry] = mutable.HashMap.empty,
    binds: List[(String, Term)] = Nil
  )

  /** Parse result: (parsed term, new state) or failure */
  type ParseResult = (Option[(Term, ParseState)], mutable.HashMap[(Int, String), MemoEntry])

  //============================================================================
  // Core Parsing Engine
  //============================================================================

  /** Find a production by name */
  def findProduction(prods: Productions, name: String): Option[GrammarExpr] =
    prods.find(_.name == name).map(_.grammar)

  /** Find all productions with the same base name and combine with alt */
  def findAllProductions(prods: Productions, name: String): Option[GrammarExpr] = {
    val matching = prods.filter(_.name == name)
    matching match {
      case Nil => None
      case p :: Nil => Some(p.grammar)
      case p :: ps => Some(ps.foldLeft(p.grammar)((acc, prod) => Alt(acc, prod.grammar)))
    }
  }

  /** Combine two terms in sequence */
  def combineSeq(t1: Term, t2: Term): Term = (t1, t2) match {
    case (Con("unit", Nil), t) => t
    case (t, Con("unit", Nil)) => t
    case (Con("seq", ts), t) => Con("seq", ts :+ t)
    case (t, Con("seq", ts)) => Con("seq", t :: ts)
    case (t1, t2) => Con("seq", List(t1, t2))
  }

  /** Wrap term in a constructor node */
  def wrapNode(name: String, t: Term): Term = Con(name, List(t))

  /** Parse grammar expression (main parsing engine)
   *  Uses fuel for termination and Packrat memoization for efficiency */
  def parseGrammar(fuel: Int, prods: Productions, g: GrammarExpr, st: ParseState): ParseResult = {
    if (fuel <= 0) return (None, st.memo)

    g match {
      case Empty => (Some((Con("unit", Nil), st)), st.memo)

      case GLit(s) =>
        st.tokens match {
          case TokLit(l) :: rest if l == s =>
            (Some((Lit(s), st.copy(tokens = rest, pos = st.pos + 1))), st.memo)
          case TokSym(l) :: rest if l == s =>
            (Some((Lit(s), st.copy(tokens = rest, pos = st.pos + 1))), st.memo)
          case TokIdent(l) :: rest if l == s =>
            (Some((Lit(s), st.copy(tokens = rest, pos = st.pos + 1))), st.memo)
          case _ => (None, st.memo)
        }

      case Ref(name) =>
        // Handle built-in token types (TOKEN.*)
        if (name.startsWith("TOKEN.")) {
          val tokType = name.drop(6)
          (tokType, st.tokens) match {
            case ("ident", TokIdent(s) :: rest) =>
              (Some((Var(s), st.copy(tokens = rest, pos = st.pos + 1))), st.memo)
            case ("string", TokLit(s) :: rest) if s.startsWith("\"") =>
              (Some((Lit(s), st.copy(tokens = rest, pos = st.pos + 1))), st.memo)
            case ("number", TokNumber(s) :: rest) =>
              (Some((Lit(s), st.copy(tokens = rest, pos = st.pos + 1))), st.memo)
            case ("sym", TokSym(s) :: rest) =>
              (Some((Var(s), st.copy(tokens = rest, pos = st.pos + 1))), st.memo)
            case _ => (None, st.memo)
          }
        } else {
          // Packrat memoization for production references
          val key = (st.pos, name)
          st.memo.get(key) match {
            case Some(entry) =>
              entry.result match {
                case Some((term, endPos, newBinds)) =>
                  val tokenCount = endPos - st.pos
                  val newTokens = st.tokens.drop(tokenCount)
                  (Some((term, st.copy(tokens = newTokens, pos = endPos, binds = newBinds))), st.memo)
                case None => (None, st.memo)
              }
            case None =>
              findAllProductions(prods, name) match {
                case Some(g2) =>
                  val (result, memo2) = parseGrammar(fuel - 1, prods, g2, st)
                  result match {
                    case Some((term, st2)) =>
                      val entry = MemoEntry(Some((term, st2.pos, st2.binds)))
                      memo2.put(key, entry)
                      (Some((term, st2.copy(memo = memo2))), memo2)
                    case None =>
                      val entry = MemoEntry(None)
                      memo2.put(key, entry)
                      (None, memo2)
                  }
                case None => (None, st.memo)
              }
          }
        }

      case Seq(g1, g2) =>
        val (r1, memo1) = parseGrammar(fuel - 1, prods, g1, st)
        r1 match {
          case Some((t1, st1)) =>
            val st1p = st1.copy(memo = memo1)
            val (r2, memo2) = parseGrammar(fuel - 1, prods, g2, st1p)
            r2 match {
              case Some((t2, st2)) => (Some((combineSeq(t1, t2), st2)), memo2)
              case None => (None, memo2)
            }
          case None => (None, memo1)
        }

      case Alt(g1, g2) =>
        val (r1, memo1) = parseGrammar(fuel - 1, prods, g1, st)
        r1 match {
          case Some(result) => (Some(result), memo1)
          case None =>
            val st2 = st.copy(memo = memo1)
            parseGrammar(fuel - 1, prods, g2, st2)
        }

      case Star(g2) =>
        def loop(acc: List[Term], st: ParseState, memo: mutable.HashMap[(Int, String), MemoEntry], loopFuel: Int): ParseResult = {
          if (loopFuel <= 0) {
            val result = if (acc.isEmpty) Con("unit", Nil) else Con("seq", acc.reverse)
            (Some((result, st)), memo)
          } else {
            val st2 = st.copy(memo = memo)
            val (r, memo2) = parseGrammar(fuel - 1, prods, g2, st2)
            r match {
              case Some((t, st3)) => loop(t :: acc, st3, memo2, loopFuel - 1)
              case None =>
                val result = if (acc.isEmpty) Con("unit", Nil) else Con("seq", acc.reverse)
                (Some((result, st.copy(memo = memo2))), memo2)
            }
          }
        }
        loop(Nil, st, st.memo, fuel - 1)

      case Plus(g2) =>
        val (first, memo1) = parseGrammar(fuel - 1, prods, g2, st)
        first match {
          case None => (None, memo1)
          case Some((t, st2)) =>
            val (rest, memo2) = parseGrammar(fuel - 1, prods, Star(g2), st2.copy(memo = memo1))
            rest match {
              case Some((Con("unit", Nil), st3)) => (Some((t, st3)), memo2)
              case Some((ts, st3)) => (Some((combineSeq(t, ts), st3)), memo2)
              case None => (Some((t, st2)), memo2)
            }
        }

      case Opt(g2) =>
        val (r, memo2) = parseGrammar(fuel - 1, prods, g2, st)
        r match {
          case Some(result) => (Some(result), memo2)
          case None => (Some((Con("unit", Nil), st.copy(memo = memo2))), memo2)
        }

      case Node(name, g2) =>
        val (r, memo2) = parseGrammar(fuel - 1, prods, g2, st)
        r match {
          case Some((t, st2)) => (Some((wrapNode(name, t), st2)), memo2)
          case None => (None, memo2)
        }
    }
  }

  //============================================================================
  // IO Operations
  //============================================================================

  /** Read file contents */
  def readFile(path: String): String =
    scala.io.Source.fromFile(path).mkString

  /** Write file contents */
  def writeFile(path: String, content: String): Unit = {
    val pw = new java.io.PrintWriter(path)
    try pw.write(content)
    finally pw.close()
  }

  /** Check if file exists */
  def fileExists(path: String): Boolean =
    new java.io.File(path).exists()

  //============================================================================
  // Term Utilities
  //============================================================================

  /** Pattern match helper */
  def matchPattern(pattern: Term, t: Term): Option[List[(String, Term)]] = (pattern, t) match {
    case (Var(x), t) => Some(List((x, t)))
    case (Lit(s1), Lit(s2)) if s1 == s2 => Some(Nil)
    case (Con(n1, args1), Con(n2, args2)) if n1 == n2 && args1.length == args2.length =>
      val results = args1.zip(args2).map { case (p, t) => matchPattern(p, t) }
      if (results.forall(_.isDefined)) Some(results.flatMap(_.get))
      else None
    case _ => None
  }

  /** Substitute bindings in a term */
  def substitute(binds: List[(String, Term)], t: Term): Term = t match {
    case Var(x) => binds.find(_._1 == x).map(_._2).getOrElse(t)
    case Lit(s) => Lit(s)
    case Con(n, args) => Con(n, args.map(substitute(binds, _)))
  }

  /** Apply a single rewrite rule */
  def applyRule(rule: Rule, t: Term): Option[Term] =
    matchPattern(rule.lhs, t).map(binds => substitute(binds, rule.rhs))

  /** Normalize term using rewrite rules */
  def normalize(rules: List[Rule], t: Term, fuel: Int = 1000): Term = {
    if (fuel <= 0) return t

    // Try each rule
    rules.flatMap(r => applyRule(r, t)).headOption match {
      case Some(t2) => normalize(rules, t2, fuel - 1)
      case None =>
        // Recursively normalize children
        t match {
          case Con(n, args) => Con(n, args.map(normalize(rules, _, fuel - 1)))
          case other => other
        }
    }
  }
}

File: src/Runtime/Rust/runtime.rs
//! Lego Runtime Library for Rust
//!
//! This module provides the core runtime infrastructure for Lego-generated Rust code.
//! It includes:
//! - Core types (Term, GrammarExpr, etc.)
//! - Parsing engine (parse_grammar, lex_grammar)
//! - IO operations (read_file, write_file)
//! - Memoization (HashMap-based Packrat)
//!
//! Generated Rosetta code should: use lego_runtime::*;

use std::collections::HashMap;
use std::fs;
use std::path::Path;

//============================================================================
// Core Types
//============================================================================

/// The universal Term type for AST representation
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Term {
    Var(String),
    Lit(String),
    Con(String, Vec<Term>),
}

impl Term {
    pub fn var(s: impl Into<String>) -> Self {
        Term::Var(s.into())
    }

    pub fn lit(s: impl Into<String>) -> Self {
        Term::Lit(s.into())
    }

    pub fn con(name: impl Into<String>, args: Vec<Term>) -> Self {
        Term::Con(name.into(), args)
    }

    pub fn unit() -> Self {
        Term::Con("unit".into(), vec![])
    }
}

/// Grammar expression algebra
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum GrammarExpr {
    Empty,
    Lit(String),
    Ref(String),
    Seq(Box<GrammarExpr>, Box<GrammarExpr>),
    Alt(Box<GrammarExpr>, Box<GrammarExpr>),
    Star(Box<GrammarExpr>),
    Plus(Box<GrammarExpr>),
    Opt(Box<GrammarExpr>),
    Node(String, Box<GrammarExpr>),
}

impl GrammarExpr {
    pub fn seq(g1: GrammarExpr, g2: GrammarExpr) -> Self {
        GrammarExpr::Seq(Box::new(g1), Box::new(g2))
    }

    pub fn alt(g1: GrammarExpr, g2: GrammarExpr) -> Self {
        GrammarExpr::Alt(Box::new(g1), Box::new(g2))
    }

    pub fn star(g: GrammarExpr) -> Self {
        GrammarExpr::Star(Box::new(g))
    }

    pub fn plus(g: GrammarExpr) -> Self {
        GrammarExpr::Plus(Box::new(g))
    }

    pub fn opt(g: GrammarExpr) -> Self {
        GrammarExpr::Opt(Box::new(g))
    }

    pub fn node(name: impl Into<String>, g: GrammarExpr) -> Self {
        GrammarExpr::Node(name.into(), Box::new(g))
    }
}

/// Grammar production: (name, grammar, annotation)
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Production {
    pub name: String,
    pub grammar: GrammarExpr,
    pub annotation: String,
}

impl Production {
    pub fn new(name: impl Into<String>, grammar: GrammarExpr) -> Self {
        Production {
            name: name.into(),
            grammar,
            annotation: String::new(),
        }
    }

    pub fn with_annotation(name: impl Into<String>, grammar: GrammarExpr, annotation: impl Into<String>) -> Self {
        Production {
            name: name.into(),
            grammar,
            annotation: annotation.into(),
        }
    }
}

/// Production list type
pub type Productions = Vec<Production>;

/// Rewrite rule: name, lhs pattern, rhs template
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Rule {
    pub name: String,
    pub lhs: Term,
    pub rhs: Term,
}

impl Rule {
    pub fn new(name: impl Into<String>, lhs: Term, rhs: Term) -> Self {
        Rule {
            name: name.into(),
            lhs,
            rhs,
        }
    }
}

//============================================================================
// Token Types
//============================================================================

/// Token types for lexing
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Token {
    Lit(String),      // String literal "..."
    Ident(String),    // Identifier
    Sym(String),      // Symbol/operator
    Number(String),   // Numeric literal
}

impl std::fmt::Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Token::Lit(s) => write!(f, "\"{}\"", s),
            Token::Ident(s) => write!(f, "{}", s),
            Token::Sym(s) => write!(f, "{}", s),
            Token::Number(s) => write!(f, "{}", s),
        }
    }
}

//============================================================================
// Parse State
//============================================================================

/// Packrat memoization entry
#[derive(Debug, Clone)]
pub struct MemoEntry {
    pub result: Option<(Term, usize, Vec<(String, Term)>)>,
}

/// Parse state
#[derive(Debug, Clone)]
pub struct ParseState {
    pub tokens: Vec<Token>,
    pub pos: usize,
    pub memo: HashMap<(usize, String), MemoEntry>,
    pub binds: Vec<(String, Term)>,
}

impl ParseState {
    pub fn new(tokens: Vec<Token>) -> Self {
        ParseState {
            tokens,
            pos: 0,
            memo: HashMap::with_capacity(256),
            binds: Vec::new(),
        }
    }

    pub fn current_token(&self) -> Option<&Token> {
        self.tokens.get(self.pos)
    }

    pub fn advance(&mut self) {
        if self.pos < self.tokens.len() {
            self.pos += 1;
        }
    }
}

/// Parse result: (parsed term, new state) or failure
pub type ParseResult = (Option<(Term, ParseState)>, HashMap<(usize, String), MemoEntry>);

//============================================================================
// Core Parsing Engine
//============================================================================

/// Find a production by name
pub fn find_production(prods: &Productions, name: &str) -> Option<GrammarExpr> {
    prods.iter()
        .find(|p| p.name == name)
        .map(|p| p.grammar.clone())
}

/// Find all productions with the same base name and combine with alt
pub fn find_all_productions(prods: &Productions, name: &str) -> Option<GrammarExpr> {
    let matching: Vec<_> = prods.iter()
        .filter(|p| p.name == name)
        .collect();

    match matching.as_slice() {
        [] => None,
        [p] => Some(p.grammar.clone()),
        [p, ps @ ..] => {
            let mut result = p.grammar.clone();
            for prod in ps {
                result = GrammarExpr::alt(result, prod.grammar.clone());
            }
            Some(result)
        }
    }
}

/// Combine two terms in sequence
pub fn combine_seq(t1: Term, t2: Term) -> Term {
    match (&t1, &t2) {
        (Term::Con(n, args), _) if n == "unit" && args.is_empty() => t2,
        (_, Term::Con(n, args)) if n == "unit" && args.is_empty() => t1,
        (Term::Con(n, args), _) if n == "seq" => {
            let mut new_args = args.clone();
            new_args.push(t2);
            Term::Con("seq".into(), new_args)
        }
        (_, Term::Con(n, args)) if n == "seq" => {
            let mut new_args = vec![t1];
            new_args.extend(args.clone());
            Term::Con("seq".into(), new_args)
        }
        _ => Term::Con("seq".into(), vec![t1, t2]),
    }
}

/// Wrap term in a constructor node
pub fn wrap_node(name: &str, t: Term) -> Term {
    Term::Con(name.into(), vec![t])
}

/// Parse grammar expression (main parsing engine)
/// Uses fuel for termination and Packrat memoization for efficiency
pub fn parse_grammar(
    fuel: usize,
    prods: &Productions,
    g: &GrammarExpr,
    st: ParseState,
) -> ParseResult {
    if fuel == 0 {
        return (None, st.memo);
    }

    match g {
        GrammarExpr::Empty => (Some((Term::unit(), st.clone())), st.memo),

        GrammarExpr::Lit(s) => {
            if let Some(tok) = st.tokens.get(st.pos) {
                let matches = match tok {
                    Token::Lit(l) => l == s,
                    Token::Sym(l) => l == s,
                    Token::Ident(l) => l == s,
                    _ => false,
                };
                if matches {
                    let mut new_st = st.clone();
                    new_st.pos += 1;
                    (Some((Term::Lit(s.clone()), new_st)), st.memo)
                } else {
                    (None, st.memo)
                }
            } else {
                (None, st.memo)
            }
        }

        GrammarExpr::Ref(name) => {
            // Handle built-in token types (TOKEN.*)
            if name.starts_with("TOKEN.") {
                let tok_type = &name[6..];
                if let Some(tok) = st.tokens.get(st.pos) {
                    let result = match (tok_type, tok) {
                        ("ident", Token::Ident(s)) => Some(Term::Var(s.clone())),
                        ("string", Token::Lit(s)) if s.starts_with('"') => Some(Term::Lit(s.clone())),
                        ("number", Token::Number(s)) => Some(Term::Lit(s.clone())),
                        ("sym", Token::Sym(s)) => Some(Term::Var(s.clone())),
                        _ => None,
                    };
                    if let Some(term) = result {
                        let mut new_st = st.clone();
                        new_st.pos += 1;
                        return (Some((term, new_st)), st.memo);
                    }
                }
                return (None, st.memo);
            }

            // Packrat memoization for production references
            let key = (st.pos, name.clone());
            if let Some(entry) = st.memo.get(&key) {
                if let Some((term, end_pos, new_binds)) = &entry.result {
                    let token_count = end_pos - st.pos;
                    let mut new_st = st.clone();
                    new_st.pos = *end_pos;
                    new_st.binds = new_binds.clone();
                    return (Some((term.clone(), new_st)), st.memo);
                } else {
                    return (None, st.memo);
                }
            }

            if let Some(g2) = find_all_productions(prods, name) {
                let (result, mut memo) = parse_grammar(fuel - 1, prods, &g2, st.clone());
                match result {
                    Some((term, st2)) => {
                        let entry = MemoEntry {
                            result: Some((term.clone(), st2.pos, st2.binds.clone())),
                        };
                        memo.insert(key, entry);
                        let mut new_st = st2;
                        new_st.memo = memo.clone();
                        (Some((term, new_st)), memo)
                    }
                    None => {
                        let entry = MemoEntry { result: None };
                        memo.insert(key, entry);
                        (None, memo)
                    }
                }
            } else {
                (None, st.memo)
            }
        }

        GrammarExpr::Seq(g1, g2) => {
            let (r1, memo1) = parse_grammar(fuel - 1, prods, g1, st);
            match r1 {
                Some((t1, mut st1)) => {
                    st1.memo = memo1;
                    let (r2, memo2) = parse_grammar(fuel - 1, prods, g2, st1);
                    match r2 {
                        Some((t2, st2)) => (Some((combine_seq(t1, t2), st2)), memo2),
                        None => (None, memo2),
                    }
                }
                None => (None, memo1),
            }
        }

        GrammarExpr::Alt(g1, g2) => {
            let (r1, memo1) = parse_grammar(fuel - 1, prods, g1, st.clone());
            match r1 {
                Some(result) => (Some(result), memo1),
                None => {
                    let mut st2 = st;
                    st2.memo = memo1;
                    parse_grammar(fuel - 1, prods, g2, st2)
                }
            }
        }

        GrammarExpr::Star(g2) => {
            let mut acc = Vec::new();
            let mut curr_st = st;
            let mut curr_memo = curr_st.memo.clone();
            let mut loop_fuel = fuel - 1;

            while loop_fuel > 0 {
                curr_st.memo = curr_memo.clone();
                let (r, memo2) = parse_grammar(fuel - 1, prods, g2, curr_st.clone());
                match r {
                    Some((t, st2)) => {
                        acc.push(t);
                        curr_st = st2;
                        curr_memo = memo2;
                        loop_fuel -= 1;
                    }
                    None => {
                        curr_memo = memo2;
                        break;
                    }
                }
            }

            let result = if acc.is_empty() {
                Term::unit()
            } else {
                Term::Con("seq".into(), acc)
            };
            curr_st.memo = curr_memo.clone();
            (Some((result, curr_st)), curr_memo)
        }

        GrammarExpr::Plus(g2) => {
            let (first, memo1) = parse_grammar(fuel - 1, prods, g2, st);
            match first {
                None => (None, memo1),
                Some((t, mut st2)) => {
                    st2.memo = memo1;
                    let (rest, memo2) = parse_grammar(fuel - 1, prods, &GrammarExpr::star((**g2).clone()), st2);
                    match rest {
                        Some((Term::Con(n, args), st3)) if n == "unit" && args.is_empty() => {
                            (Some((t, st3)), memo2)
                        }
                        Some((ts, st3)) => (Some((combine_seq(t, ts), st3)), memo2),
                        None => (Some((t, st2.clone())), memo2),
                    }
                }
            }
        }

        GrammarExpr::Opt(g2) => {
            let (r, memo2) = parse_grammar(fuel - 1, prods, g2, st.clone());
            match r {
                Some(result) => (Some(result), memo2),
                None => {
                    let mut new_st = st;
                    new_st.memo = memo2.clone();
                    (Some((Term::unit(), new_st)), memo2)
                }
            }
        }

        GrammarExpr::Node(name, g2) => {
            let (r, memo2) = parse_grammar(fuel - 1, prods, g2, st);
            match r {
                Some((t, st2)) => (Some((wrap_node(name, t), st2)), memo2),
                None => (None, memo2),
            }
        }
    }
}

//============================================================================
// IO Operations
//============================================================================

/// Read file contents
pub fn read_file<P: AsRef<Path>>(path: P) -> std::io::Result<String> {
    fs::read_to_string(path)
}

/// Write file contents
pub fn write_file<P: AsRef<Path>>(path: P, content: &str) -> std::io::Result<()> {
    fs::write(path, content)
}

/// Check if file exists
pub fn file_exists<P: AsRef<Path>>(path: P) -> bool {
    path.as_ref().exists()
}

//============================================================================
// Term Utilities
//============================================================================

/// Pattern match helper
pub fn match_pattern(pattern: &Term, t: &Term) -> Option<Vec<(String, Term)>> {
    match (pattern, t) {
        (Term::Var(x), t) => Some(vec![(x.clone(), t.clone())]),
        (Term::Lit(s1), Term::Lit(s2)) if s1 == s2 => Some(vec![]),
        (Term::Con(n1, args1), Term::Con(n2, args2))
            if n1 == n2 && args1.len() == args2.len() =>
        {
            let mut all_binds = Vec::new();
            for (p, t) in args1.iter().zip(args2.iter()) {
                match match_pattern(p, t) {
                    Some(binds) => all_binds.extend(binds),
                    None => return None,
                }
            }
            Some(all_binds)
        }
        _ => None,
    }
}

/// Substitute bindings in a term
pub fn substitute(binds: &[(String, Term)], t: &Term) -> Term {
    match t {
        Term::Var(x) => {
            binds.iter()
                .find(|(name, _)| name == x)
                .map(|(_, val)| val.clone())
                .unwrap_or_else(|| t.clone())
        }
        Term::Lit(s) => Term::Lit(s.clone()),
        Term::Con(n, args) => {
            Term::Con(n.clone(), args.iter().map(|a| substitute(binds, a)).collect())
        }
    }
}

/// Apply a single rewrite rule
pub fn apply_rule(rule: &Rule, t: &Term) -> Option<Term> {
    match_pattern(&rule.lhs, t).map(|binds| substitute(&binds, &rule.rhs))
}

/// Normalize term using rewrite rules
pub fn normalize(rules: &[Rule], t: &Term, fuel: usize) -> Term {
    if fuel == 0 {
        return t.clone();
    }

    // Try each rule
    for rule in rules {
        if let Some(t2) = apply_rule(rule, t) {
            return normalize(rules, &t2, fuel - 1);
        }
    }

    // Recursively normalize children
    match t {
        Term::Con(n, args) => {
            Term::Con(n.clone(), args.iter().map(|a| normalize(rules, a, fuel - 1)).collect())
        }
        other => other.clone(),
    }
}

//============================================================================
// Tests
//============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_term_construction() {
        let t = Term::con("add", vec![Term::lit("1"), Term::lit("2")]);
        assert_eq!(t, Term::Con("add".into(), vec![Term::Lit("1".into()), Term::Lit("2".into())]));
    }

    #[test]
    fn test_pattern_match() {
        let pattern = Term::con("add", vec![Term::var("x"), Term::var("y")]);
        let term = Term::con("add", vec![Term::lit("1"), Term::lit("2")]);

        let result = match_pattern(&pattern, &term);
        assert!(result.is_some());

        let binds = result.unwrap();
        assert_eq!(binds.len(), 2);
    }

    #[test]
    fn test_substitute() {
        let binds = vec![
            ("x".into(), Term::lit("1")),
            ("y".into(), Term::lit("2")),
        ];
        let template = Term::con("result", vec![Term::var("x"), Term::var("y")]);

        let result = substitute(&binds, &template);
        assert_eq!(result, Term::con("result", vec![Term::lit("1"), Term::lit("2")]));
    }
}
